aiida_orca/workchains/base.py

```diff
@@ -46,5 +51,9 @@ def report_error_handled(self, calculation, action):
         self.report('{}<{}> failed with exit status {}: {}'.format(*arguments))
         self.report('Action taken: {}'.format(action))
 
-
-#EOF
+    @process_handler(priority=600)
+    def handle_unrecoverable_failure(self, calculation):
+        """Handle calculations with an exit status below 400 which are unrecoverable, so abort the work chain."""
+        if calculation.is_failed and calculation.exit_status < 400:
```

Comment:
Hi @sphuber,

I am trying to implement the `process_handler` for the `OrcaBaseWorkChain` which subclasses `BaseRestartWorkChain`.  Without the process handler, the `BaseRestartWorkChain` needlessly restarts the calculation even if we report unrecoverable error (e.g. user provided bad input).

I've essentially just copied this code from `aiida-quantumespresso` plugin but I am struggling to understand why it was implemented this way. Specifically, the [exit code guidelines](https://aiida.readthedocs.io/projects/aiida-core/en/latest/topics/processes/usage.html?highlight=exit%20codes#exit-code-conventions) in AiiDA docs say

>         0 - 99: Reserved for internal use by aiida-core
>        100 - 199: Reserved for errors parsed from scheduler output of calculation jobs (note: this is not yet implemented)
>        200 - 299: Suggested to be used for process input validation errors
>        300 - 399: Suggested for critical process errors

However, the if statement in this handler includes **all** error codes below 400, which possibly includes aiida-core errors, scheduler errors etc. I'd think that at least some of these errors might be recoverable, and thus should not be covered by this process_handler that returns non-zero (and thus ends the `BaseRestartWorkChain` iterations). Am I missing something?

Reply:
You are not really missing something. Note that I implemented this handler before there was scheduler error parsing (as indicated by the comment in the row of 100 - 199 exit codes). Nevertheless, now that there are some errors being parsed (by the `SlurmScheduler` plugin at least, not aware of any others doing this as well) these are all errors that would most likely render the calculation non-restartable, at least in the case of `PwCalculation`. These are out-of-walltime, out-of-memory, and node failures. So to me the logic still seems ok.

That being said, I have actually been considering to disable the handler by default. The reason I added it is because alsmost always (during my testing) these errors really were unrecoverable and so the default second retry after an unhandled error, would always result in a second failure. However, there are hard failures that could be transient, for example the node failure (not always the case, depends on the reason), and a restart may just work the second time around.

In the end though, you can implement whatever logic you feel makes most sense for ORCA. If you think there are exit codes that can be recoverable (even if they are below 400, which in the end is just a very loose convention and is not enforced anywhere in `aiida-core`) you can just code that in your handlers.

Reply:
Thanks! 

> out-of-walltime, out-of-memory

Well, aren't these possibly restartable by changing the queue settings? (e.g. submitting to a queue that allows longer wall-time, or has nodes with larger memory).

I guess my main point is that I as a plugin developer cannot possibly know all errors that might come from aiida-core or schedulers. What would make sense to me is if the plugin handled errors between 200-400, since those I have control over. Errors between 1-200 could be handled by `BaseRestartWorkChain`. But at the moment it looks like `BaseRestartWorkChain` doesn't handle any errors, so I have to assume none of the errors are restartable by default, which essentially reverses the default logic of BaseRestartWorkChain. :shrug:  Since we currently don't have any restartable errors for ORCA, it looks like the the we would never ever restart with this logic, defeating the whole purpose?

Which makes me think why is the default logic of BaseRestartWorkChain to restart, when the vast majority of errors are probably non-restartable, especially without any extra custom logic? Are there any examples from aiida-core or schedulers which are known to be transient and thus it makes sense to restart. If yes, I'd argue they should be added to the BaseRestartWorkChain, otherwise I will need to add them here...

Reply:
The `BaseRestartWorkChain` does indeed not handle any errors by default, but this is really by design, because how could it? The base implementation doesn't know what calculation is going to be run and so cannot know what errors to expect or how to fix them. Of course there are the "generic" ones, but as said before, they didn't exist when I developed the `BaseRestartWorkChain`, and even now, I am not sure if we can have generic handlers for these errors.

> Well, aren't these possibly restartable by changing the queue settings? (e.g. submitting to a queue that allows longer wall-time, or has nodes with larger memory).

This was a problem that I cracked my head over for a long time when I first developed the `BaseRestartWorkChain` for the `PwCalculation`. I concluded that I would prefer the base workchain not to arbitrarily increase the number of resources. The reason being that the user might not want more resources to be used. You could say that the workchain can increase a bit but not too much, but where is the limit? So I decided that I prefer this to be dealt with differnently, higher up the stack. This is a choice and you could of course decide to do it differently and have your implementation deal with these errors and automatically increase resources.

Ultimately, the `BaseRestartWorkChain` is some scaffolding to make it easier to implement restart logic around a `CalcJob` plugin, _if_ that has some known restartable errors. I agree completely that if ORCA doesn't have any failure modes that are restartable, then implementing the base workchain is pointless. But no one force you to use it.

Reply:
@sphuber thanks so much for the write-up, really made things clearer.

> Ultimately, the BaseRestartWorkChain is some scaffolding to make it easier to implement restart logic around a CalcJob plugin, if that has some known restartable errors.

Right, this is a key sentence, I definitely had some misconceptions here. Now it makes more sense. 

In this PR I kept the workchain and implemented the error handlers since it is likely we will have restartable errors in the future (e.g. SCF convergence).